File: python/trees/serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """

        def dfs(node):
            if not node:
                return "N"

            return str(node.val) + " " + dfs(node.left) + " " + dfs(node.right)

        return dfs(root)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        processed = data.split(" ")
        if processed[0] == "N":
            return None

        final = TreeNode(int(processed[0]))

        def construct(node, index):
            current = index
            if processed[current] != "N":
                node.left = TreeNode(int(processed[current]))
                current = construct(node.left, current + 1)
            else:
                current += 1

            if processed[current] != "N":
                node.right = TreeNode(int(processed[current]))
                current = construct(node.right, current + 1)
            else:
                current += 1

            return current

        construct(final, 1)
        return final

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: c++/trees/serialize-and-deserialize-binary-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        std::string encoding = encode(root);
        return encoding.substr(0, encoding.size()-1);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        std::vector<std::string> tokens;
        std::stringstream ss(data);
        std::string token;

        while (ss >> token) {
            tokens.push_back(token);
        }

        std::reverse(tokens.begin(), tokens.end());

        return construct(tokens);
    }

private:
    string encode(TreeNode* node) {
        if (!node) {
            return "None ";
        }

        std::string encoding = std::to_string(node->val) + " ";
        encoding += encode(node->left);
        encoding += encode(node->right);

        return encoding;
    }

    TreeNode* construct(std::vector<std::string>& tokens) {
        if (tokens.back() == "None") {
            tokens.pop_back();
            return nullptr;
        }

        TreeNode* node = new TreeNode(std::stoi(tokens.back()));
        tokens.pop_back();
        node->left = construct(tokens);
        node->right = construct(tokens);

        return node;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));
